app/scripts/modules/core/cluster/filter/clusterFilter.model.ts

```diff
@@ -0,0 +1,114 @@
+
+  private mostRecentParams: any;
+  private asFilterModel: IFilterModel;
+
+  static get $inject() { return ['$rootScope', 'filterModelService', 'urlParser']; }
+
+  constructor(private $rootScope: IRootScopeService, private filterModelService: any, private urlParser: any) {
+    this.asFilterModel = this.filterModelService.configureFilterModel(this, filterModelConfig);
+    this.bindEvents();
+    this.asFilterModel.activate();
+  }
+
+  private isClusterState(stateName: string): boolean {
+    return stateName === 'home.applications.application.insight.clusters' ||
+      stateName === 'home.project.application.insight.clusters';
+  }
+
+  private isClusterStateOrChild(stateName: string): boolean {
+    return this.isClusterState(stateName) || this.isChildState(stateName);
+  }
+
+  private isChildState(stateName: string): boolean {
+    return stateName.includes('clusters.');
+  }
+
+  private movingToClusterState(toState: State): boolean {
+    return this.isClusterStateOrChild(toState.name);
+  }
+
+  private movingFromClusterState (toState: State, fromState: State): boolean {
+    return this.isClusterStateOrChild(fromState.name) && !this.isClusterStateOrChild(toState.name);
+  }
+
+  private fromApplicationListState(fromState: State): boolean {
+    return fromState.name === 'home.applications';
+  }
+
+  private shouldRouteToSavedState(toParams: StateParams, fromState: State): boolean {
+    return this.asFilterModel.hasSavedState(toParams) && !this.isClusterStateOrChild(fromState.name);
+  }
+
+  private handleStateChangeStart(_event: IAngularEvent, toState: State, _toParams: StateParams, fromState: State, fromParams: StateParams): void {
+    if (this.movingFromClusterState(toState, fromState)) {
+      this.asFilterModel.saveState(fromState, fromParams, this.mostRecentParams);
+    }
+  };
+
+  private bindEvents(): void {
+    // WHY??? Because, when the stateChangeStart event fires, the $location.search() will return whatever the query
+    // params are on the route we are going to, so if the user is using the back button, for example, to go to the
+    // Infrastructure page with a search already entered, we'll pick up whatever search was entered there, and if we
+    // come back to this application's clusters view, we'll get whatever that search was.
+    this.$rootScope.$on('$locationChangeStart', (_event: IAngularEvent, toUrl: string, fromUrl: string) => {
+      const [oldBase, oldQuery] = fromUrl.split('?'),
+        [newBase, newQuery] = toUrl.split('?');
+
+      if (oldBase === newBase) {
+        this.mostRecentParams = newQuery ? this.urlParser.parseQueryString(newQuery) : {};
+      } else {
+        this.mostRecentParams = oldQuery ? this.urlParser.parseQueryString(oldQuery) : {};
+      }
+    });
+
+    this.$rootScope.$on('$stateChangeStart', this.handleStateChangeStart);
+
+    this.$rootScope.$on('$stateChangeSuccess', (_event: IAngularEvent, toState: State, toParams: StateParams, fromState: State) => {
```

Comment:
Do these `$rootScope.$on('$stateChange*...` listeners still make sense in `ui-router@1.0` world?